geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImpl.java

```diff
@@ -0,0 +1,176 @@
+    ALLOW_UNTRUSTED_METHOD_INVOCATION = Boolean.parseBoolean(
+        System.getProperty(GEMFIRE_PREFIX + "QueryService.allowUntrustedMethodInvocation"));
+  }
+
+  public static MethodInvocationAuthorizer getNoOpAuthorizer() {
+    // A no-op authorizer, allow method invocation
+    return ((Method m, Object t) -> true);
+  }
+
+  @Override
+  public boolean init(Cache cache) {
+    if (cache == null) {
+      throw new IllegalArgumentException("cache must not be null");
+    }
+
+    if (System
+        .getProperty(GEMFIRE_PREFIX + "QueryService.allowUntrustedMethodInvocation") != null) {
+      logger.warn(DEPRECATION_WARNING);
+    }
+
+    if (isSecurityDisabled((InternalCache) cache) || ALLOW_UNTRUSTED_METHOD_INVOCATION) {
+      this.authorizer = getNoOpAuthorizer();
+    } else {
+      this.authorizer = new RestrictedMethodAuthorizer(cache);
+    }
+    return true;
+  }
+
+  @Override
+  public Class<? extends CacheService> getInterface() {
+    return QueryConfigurationService.class;
+  }
+
+  @Override
+  public CacheServiceMBeanBase getMBean() {
+    return null;
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public MethodInvocationAuthorizer getMethodAuthorizer() {
+    return authorizer;
+  }
+
+  @Override
+  public void updateMethodAuthorizer(Cache cache, QueryMethodAuthorizerCreation creation) {
+    updateMethodAuthorizer(cache, creation.getClassName(), creation.getParameters());
+  }
+
+  @Override
+  public void updateMethodAuthorizer(Cache cache, String className, Set<String> parameters) {
+    if (isSecurityDisabled((InternalCache) cache) || ALLOW_UNTRUSTED_METHOD_INVOCATION) {
+      return;
+    }
+
+    if (className == null) {
+      logError(UPDATE_ERROR_MESSAGE + NULL_CLASS_NAME + AUTHORIZER_NOT_UPDATED);
+      return;
+    }
+
+    if (className.equals(RestrictedMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new RestrictedMethodAuthorizer(cache);
+    } else if (className.equals(UnrestrictedMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new UnrestrictedMethodAuthorizer(cache);
+    } else if (className.equals(JavaBeanAccessorMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new JavaBeanAccessorMethodAuthorizer(cache, parameters);
+    } else if (className.equals(RegExMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new RegExMethodAuthorizer(cache, parameters);
+    } else {
+      try {
+        @SuppressWarnings("unchecked")
+        Class<MethodInvocationAuthorizer> userClass =
+            (Class<MethodInvocationAuthorizer>) ClassPathLoader.getLatest().forName(className);
+        Constructor<MethodInvocationAuthorizer> userConstructor =
+            userClass.getDeclaredConstructor(Cache.class, Set.class);
+        this.authorizer = userConstructor.newInstance(cache, parameters);
+      } catch (Exception e) {
+        logErrorMessage(className, e);
+      }
+    }
+  }
```

Comment:
Random thought: should we use a lock or something to synchronize these methods?, what happens if multiple threads try to update/read the authorizer at the same time?.

Reply:
we probably need to solve this on the distributed level as well.

Reply:
That's a valid concern. Adding some kind of synchronization on the class and/or some distributed lock to prevent the possible future case of multiple gfsh commands leading to a mismatch in authorizer configuration would be a good idea.

Reply:
We looked into adding a Dlock but realized that putting a Dlock in this code would only synchronize the actual assignment of the authorizer on each member. What we want to do is protect multiple processes attempting to assign this authorizer across the cluster from running over each other. To do that we need to add a lock to the places which trigger this update to be called across the distributed system. Since we can't block the Cache Configuration process the only place we could add such a lock would be in the gfsh command which will be added in the following story for this feature.